### PR TITLE
test: allow producing block on top of arbitrary prev block

### PR DESCRIPTION
**Pull request by Longarithm**
For stateless validation, chunk execution will be delayed until next block is processed: #9982. This impacts several tests assuming that block processing includes chunk processing as well. For these tests, we need to produce and process one more block to get execution results like ChunkExtra and ExecutionOutcome.

To allow production of longer forks, I want to extend client API by `produce_block_on` which can produce a block not just on top of head, but on top of any existing block. As this block isn't immediately saved or processed, it even doesn't break any guarantees.

## Testing

Impacted tests should still pass. Later stateless validation PRs will rely on it.

### CODE REVIEW

**Comment by wacban** [comment]
Ah, finally, near gets proper support for forks, a feature long awaited by the community :)

File: chain/client/src/client.rs [hunk 1]
```diff
@@ -477,20 +477,19 @@ impl Client {
 
     fn should_reschedule_block(
```
Comment (wacban) [inline chain/client/src/client.rs:478]: While you're at it, can you add some comments please?

File: chain/client/src/client.rs [hunk 2]
```diff
@@ -565,31 +564,40 @@ impl Client {
...
 
+        self.produce_block_on(height, head.last_block_hash)
+    }
+
+    /// Produce block for given `height` on top of block `prev_hash`.
+    /// Should be called either from `produce_block` or in tests.
+    pub fn produce_block_on(
+        &mut self,
+        height: BlockHeight,
+        prev_hash: CryptoHash,
```
Comment (wacban) [inline chain/client/src/client.rs:586]: So just for my understanding, height and prev_hash are not tied together are they? As in height can be arbitrary number as long as it is greater than the height of the block with hash prev_hash.

**Review by wacban — APPROVED** [review]
LGTM

File: chain/client/src/client.rs (hunk 2 above)
Comment (Longarithm) [reply chain/client/src/client.rs:586]: Depends on what you mean by "tied together". Yeah, height is not necessarily height(prev_hash)+1 anymore, but prev_hash is still hash of previous block for block being produced.

File: chain/client/src/client.rs (hunk 1 above)
Comment (Longarithm) [reply chain/client/src/client.rs:478]: Will do in following PR.

**Comment by Longarithm** [comment]
To be completely clear, after that clients still will not produce blocks for arbitrary heights during normal production behaviour. It is test-only feature.

**Pull request merged**